url/url.cpp

```diff
@@ -51,4 +101,1341 @@ std::string blob_url_create(Origin const &origin) {
     return result;
 }
 
+void UrlParser::validation_error(ValidationError err) {
+    spdlog::debug("url: InputPos: {}, ParserState: {}, Validation Error: {} {}",
+            current_pos(),
+            std::to_underlying(state_),
+            std::to_underlying(err),
+            validation_error_str.at(err));
+}
+
+// https://url.spec.whatwg.org/#concept-url-parser
+std::optional<Url> UrlParser::parse(std::string input, std::optional<Url> base, std::optional<std::string> encoding) {
+    std::optional<Url> url = parse_basic(input, base, encoding, std::nullopt, std::nullopt);
+
+    if (url.has_value() && url->scheme == "blob") {
+        // TODO(zero-one): Resolve blob URL
+    }
+
+    return url;
+}
+
+// https://url.spec.whatwg.org/#concept-basic-url-parser
+std::optional<Url> UrlParser::parse_basic(std::string input,
+        std::optional<Url> base,
+        [[maybe_unused]] std::optional<std::string> encoding,
+        std::optional<Url> url,
+        std::optional<ParserState> state_override) {
+    base_ = base;
+    state_override_ = state_override;
+
+    if (!url.has_value()) {
+        // Set url to a new URL
+        url_ = Url();
+        url_.path = std::vector<std::string>{};
+
+        bool leading_trailing_c0 = false;
+
+        while (util::is_c0_or_space(input.front())) {
```

Comment:
I think an input of just a space (or any c0 characters) would explode here?

Reply:
`The behavior is undefined if empty() == true`. whoops